File: internal/models/workspace.go
// agency-gateway/internal/models/workspace.go
package models

import "fmt"

// WorkspaceBase defines the base container image and runtime settings.
type WorkspaceBase struct {
	Image      string `yaml:"image" validate:"required"`
	User       string `yaml:"user" default:"agent"`
	Filesystem string `yaml:"filesystem" default:"readonly-root"`
}

// WorkspaceProvides describes the capabilities exposed by a workspace.
type WorkspaceProvides struct {
	Tools   []string `yaml:"tools"`
	Network string   `yaml:"network" default:"mediated"`
}

// WorkspaceResources defines container resource limits for a workspace.
type WorkspaceResources struct {
	Memory string `yaml:"memory" default:"2GB"`
	CPU    string `yaml:"cpu" default:"1.0"`
	Tmpfs  string `yaml:"tmpfs" default:"512MB"`
}

// WorkspaceSecurity defines the security posture for a workspace container.
type WorkspaceSecurity struct {
	Capabilities    string `yaml:"capabilities" validate:"required,oneof=none" default:"none"`
	Seccomp         string `yaml:"seccomp" validate:"required,oneof=default-strict default" default:"default-strict"`
	NoNewPrivileges bool   `yaml:"no_new_privileges" default:"true"`
}

// WorkspaceConfig is the schema for workspace.yaml template files.
type WorkspaceConfig struct {
	Name      string             `yaml:"name" validate:"required"`
	Version   string             `yaml:"version" default:"1.0"`
	Base      WorkspaceBase      `yaml:"base"`
	Provides  WorkspaceProvides  `yaml:"provides"`
	Resources WorkspaceResources `yaml:"resources"`
	Security  WorkspaceSecurity  `yaml:"security"`
}

// ExtraMount defines an additional host-to-container bind mount.
type ExtraMount struct {
	Source string `yaml:"source" validate:"required"`
	Target string `yaml:"target" validate:"required"`
}

// Validate checks that both Source and Target are absolute paths.
func (e *ExtraMount) Validate() error {
	if len(e.Source) == 0 || e.Source[0] != '/' {
		return fmt.Errorf("source must be an absolute path")
	}
	if len(e.Target) == 0 || e.Target[0] != '/' {
		return fmt.Errorf("target must be an absolute path")
	}
	return nil
}

// AgentWorkspaceConfig is the schema for workspace.yaml files under agents/.
type AgentWorkspaceConfig struct {
	Version      string       `yaml:"version" default:"0.1"`
	Agent        string       `yaml:"agent" validate:"required"`
	WorkspaceRef string       `yaml:"workspace_ref" validate:"required"`
	ProjectDir   *string      `yaml:"project_dir"`
	ExtraMounts  []ExtraMount `yaml:"extra_mounts"`
}

File: images/models/workspace.py
"""Workspace configuration schema."""

from typing import Optional, Literal

from pydantic import BaseModel, ConfigDict, field_validator


class WorkspaceBase(BaseModel):
    model_config = ConfigDict(extra="forbid")
    image: str
    user: str = "agent"
    filesystem: str = "readonly-root"


class WorkspaceProvides(BaseModel):
    model_config = ConfigDict(extra="forbid")
    tools: list[str] = []
    network: str = "mediated"


class WorkspaceResources(BaseModel):
    model_config = ConfigDict(extra="forbid")
    memory: str = "2GB"
    cpu: str = "1.0"
    tmpfs: str = "512MB"


class WorkspaceSecurity(BaseModel):
    model_config = ConfigDict(extra="forbid")
    capabilities: Literal["none"] = "none"
    seccomp: Literal["default-strict", "default"] = "default-strict"
    no_new_privileges: bool = True


class WorkspaceConfig(BaseModel):
    """Schema for workspace.yaml (workspace template)."""

    model_config = ConfigDict(extra="forbid")

    name: str
    version: str = "1.0"
    base: WorkspaceBase
    provides: WorkspaceProvides = WorkspaceProvides()
    resources: WorkspaceResources = WorkspaceResources()
    security: WorkspaceSecurity = WorkspaceSecurity()


class ExtraMount(BaseModel):
    """An additional read-only bind mount for an agent workspace.

    Extra mounts are always read-only (tenet 4: access matches purpose).
    The operator configures these in workspace.yaml.
    """

    model_config = ConfigDict(extra="forbid")

    source: str  # Host path
    target: str  # Container path

    @field_validator("source")
    @classmethod
    def source_must_be_absolute(cls, v: str) -> str:
        if not v.startswith("/"):
            raise ValueError("source must be an absolute path")
        return v

    @field_validator("target")
    @classmethod
    def target_must_be_absolute(cls, v: str) -> str:
        if not v.startswith("/"):
            raise ValueError("target must be an absolute path")
        return v


class AgentWorkspaceConfig(BaseModel):
    """Schema for agent-level workspace.yaml (references a workspace template)."""

    model_config = ConfigDict(extra="forbid")

    version: str = "0.1"
    agent: str
    workspace_ref: str
    project_dir: Optional[str] = None
    extra_mounts: list[ExtraMount] = []
